pkg/controller/util/propagatedversion.go

```diff
@@ -0,0 +1,101 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+
+	fedv1a1 "github.com/marun/federation-v2/pkg/apis/federation/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+)
+
+type VersionCompareType int
+
+const (
+	ResourceVersion VersionCompareType = iota
+	Generation
+)
+
+type ComparisonHelper interface {
+	GetVersion(objectMeta *metav1.ObjectMeta) string
+	Equivalent(objMeta1, objectMeta2 *metav1.ObjectMeta) bool
+}
+
+// NewComparisonHelper instantiates and returns a Resource or Generation Helper
+// struct that implements the ComparisonHelper interface based on the version
+// comparison type passed in.
+func NewComparisonHelper(compareType VersionCompareType) (ComparisonHelper, error) {
+	switch compareType {
+	case ResourceVersion:
+		return &ResourceHelper{}, nil
+	case Generation:
+		return &GenerationHelper{}, nil
+	default:
+		err := fmt.Errorf("Unrecognized version comparison type %v", compareType)
+		runtime.HandleError(err)
+		return nil, err
+	}
+}
+
+type GenerationHelper struct{}
+
+// GetVersion returns a string containing the version in the resource's
+// ObjectMeta using the resource comparison type to perform for that
+// resource.
+func (GenerationHelper) GetVersion(objectMeta *metav1.ObjectMeta) string {
+	return strconv.FormatInt(objectMeta.Generation, 10)
+}
+
+// Equivalent returns true if both object metas passed in are equivalent, false
+// otherwise.
+func (GenerationHelper) Equivalent(obj1Meta, obj2Meta *metav1.ObjectMeta) bool {
+	return ObjectMetaEquivalent(*obj1Meta, *obj2Meta)
+}
+
+type ResourceHelper struct{}
+
+// GetVersion returns a string containing the version in the resource's
+// ObjectMeta using the resource comparison type to perform for that
+// resource.
+func (ResourceHelper) GetVersion(objectMeta *metav1.ObjectMeta) string {
+	return objectMeta.ResourceVersion
+}
+
+// Equivalent returns true (not implemented).
+func (ResourceHelper) Equivalent(obj1Meta, obj2Meta *metav1.ObjectMeta) bool {
+	return true
+}
+
+// SortClusterVersions ASCII sorts the given cluster versions slice
+// based on cluster name.
+func SortClusterVersions(versions []fedv1a1.ClusterObjectVersion) {
+	sort.Slice(versions, func(i, j int) bool {
+		return versions[i].ClusterName < versions[j].ClusterName
+	})
+}
+
+// PropagatedVersionStatusEquivalent returns true if both statuses are equal by
+// comparing Template and Override version, and their ClusterVersion slices;
+// false otherwise.
+func PropagatedVersionStatusEquivalent(pvs1, pvs2 *fedv1a1.PropagatedVersionStatus) bool {
+	return pvs1.TemplateVersion == pvs2.TemplateVersion &&
+		pvs1.OverrideVersion == pvs2.OverrideVersion &&
+		len(pvs1.ClusterVersions) == len(pvs2.ClusterVersions)
```

Comment:
There should be a comparison of the `ClusterVersions` slices with `reflect.DeepEqual`.  The fact that this omission wasn't detected in testing is reminding me that unit testing needs to start real soon.

Reply:
Yeah that's better. I knew there had to be an easier way to compare.